// !API_VERSION: 1.2

val a = <!UNSIGNED_LITERAL_WITHOUT_DECLARATIONS_ON_CLASSPATH, INT_LITERAL_OUT_OF_RANGE!>0u<!>
val b = <!UNSIGNED_LITERAL_WITHOUT_DECLARATIONS_ON_CLASSPATH, INT_LITERAL_OUT_OF_RANGE!>1uL<!>

fun foo(<!UNUSED_PARAMETER!>u<!>: <!UNRESOLVED_REFERENCE!>UInt<!>) {}