val u1 = <!UNSIGNED_LITERAL_WITHOUT_DECLARATIONS_ON_CLASSPATH, INT_LITERAL_OUT_OF_RANGE!>1u<!>
val u2 = <!UNSIGNED_LITERAL_WITHOUT_DECLARATIONS_ON_CLASSPATH, INT_LITERAL_OUT_OF_RANGE!>0xFu<!>
val u3 = <!UNSIGNED_LITERAL_WITHOUT_DECLARATIONS_ON_CLASSPATH, INT_LITERAL_OUT_OF_RANGE!>0b1u<!>